# Load the Data

In [1]:
import json
import pickle

import gensim
from gensim.models import CoherenceModel
from tqdm import tqdm_notebook

tqdm_notebook().pandas()

C:\Users\yenter\Anaconda3\lib\site-packages\gensim\utils.py:1197: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")


In [2]:
with open("processed_data/token.json") as f:
    tokens = json.load(f)

# Prepare Corpus and Dictionary

In [3]:
dictionary = gensim.corpora.Dictionary(tokens)
dictionary.save("processed_data/dictionary")

len(dictionary)

In [3]:
# dictionary = gensim.corpora.Dictionary.load('processed_data/dictionary')
dictionary.filter_extremes(no_below=20, no_above=0.5, keep_n=None)

In [5]:
corpus = [0] * len(tokens)

for i in tqdm_notebook(range(len(tokens))):
    corpus[i] = dictionary.doc2bow(tokens[i])

with open("processed_data/corpus.json", "w") as f:
    json.dump(corpus, f)
    
with open("processed_data/corpus.pkl", "wb") as f:
    pickle.dump(corpus, f)

In [6]:
# with open("processed_data/corpus.json") as f:
#     corpus = json.load(f)

# with open("processed_data/corpus.pkl", "rb") as f:
#     corpus = pickle.load(f)

# Conduct Grid Search on LDA Models

In [8]:
for passes in [1, 4]:
    for num_topics in [5, 10, 15, 20, 25, 30]:
        print("Num topics:", num_topics, ", passes:", passes)
        lda_model = gensim.models.ldamulticore.LdaMulticore(corpus=corpus,
                                                            id2word=dictionary,
                                                            num_topics=num_topics, 
                                                            random_state=1,
                                                            chunksize=2000,
                                                            passes=passes,
                                                            workers=7,
                                                            per_word_topics=True)

        coherence = CoherenceModel(model=lda_model, texts=tokens, dictionary=dictionary, coherence='c_v').get_coherence()
        print("    Coherence:", coherence)
        lda_model.save("models/lda/"+str(passes)+"-passes_" + str(num_topics) + "-topics")

Num topics: 30
    Coherence: 0.5299694323860216


| No. of topics | 1 pass | 4 passes |
| ---- | ----- | ------|
|  5 topics | 0.40667 | 0.40749 |
| 10 topics | 0.43186 | 0.45273 |
| 15 topics | 0.49410 | 0.50778 |
| 20 topics | 0.51291 | 0.53482 |
| 25 topics | 0.51477 | 0.53167 |
| 30 topics | 0.52444 | 0.52997 |



# Display Topics for Top Few Models

In [14]:
from tabulate import tabulate
from IPython.display import display, HTML, Markdown
import re

In [19]:
for num_passes, num_topics in [(4, 20), (4, 25), (4, 30), (1, 30), (1, 25), (1, 20)]:
    model = gensim.models.ldamulticore.LdaMulticore.load('models/lda/' + str(num_passes) + '-passes_' + str(num_topics) + '-topics')
    results = []
    for _, text in model.print_topics():
        results.append([re.findall('[a-z]+', x)[0] for x in text.split(' + ')])
    display(Markdown("# " + str(num_passes) + " passes, " + str(num_topics) + " topics"))
    display(HTML(tabulate(results, tablefmt="html", headers=["Term " + str(x) for x in range(1, 11)])))

# 4 passes, 20 topics

# 4 passes, 25 topics

# 4 passes, 30 topics

# 1 passes, 30 topics

# 1 passes, 25 topics

# 1 passes, 20 topics